tensorflow/python/keras/backend.py

```diff
@@ -1506,49 +1514,106 @@ def batch_dot(x, y, axes=None):
   ```python
       >>> x_batch = K.ones(shape=(32, 20, 1))
       >>> y_batch = K.ones(shape=(32, 30, 20))
-      >>> xy_batch_dot = K.batch_dot(x_batch, y_batch, axes=[1, 2])
+      >>> xy_batch_dot = K.batch_dot(x_batch, y_batch, axes=(1, 2))
       >>> K.int_shape(xy_batch_dot)
       (32, 1, 30)
   ```
   """
+  x_shape = int_shape(x)
+  y_shape = int_shape(y)
+
+  x_ndim = len(x_shape)
+  y_ndim = len(y_shape)
+
+  if x_ndim < 2 or y_ndim < 2:
+    raise ValueError('Can not do batch_dot on inputs '
+                      'with rank < 2. '
+                      'Received inputs with shapes ' +
+                      str(x_shape) + ' and ' +
+                      str(y_shape) + '.')
+
+  x_batch_size = x_shape[0]
+  y_batch_size = y_shape[0]
+
+  if x_batch_size is not None and y_batch_size is not None:
+    if x_batch_size != y_batch_size:
+        raise ValueError('Can not do batch_dot on inputs '
+                          'with different batch sizes. '
+                          'Received inputs with shapes ' +
+                          str(x_shape) + ' and ' +
+                          str(y_shape) + '.')
+
   if isinstance(axes, int):
-    axes = (axes, axes)
-  x_ndim = ndim(x)
-  y_ndim = ndim(y)
+    axes = [axes, axes]
```

Comment:
Use tuple?

Reply:
Same below.

Reply:
Has to be mutable for in-place subtraction.

Reply:
Ack. Thanks.